File: hooks/scripts/cmd/usage_awareness.go
package cmd

import (
	"encoding/json"
	"fmt"
	"net/http"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
	"strings"
	"time"

	"solon-hooks/internal/config"

	"github.com/spf13/cobra"
)

var usageAwarenessCmd = &cobra.Command{
	Use:   "usage-awareness",
	Short: "Handle UserPromptSubmit/PostToolUse usage context awareness",
	RunE:  runUsageAwareness,
}

const (
	usageCacheFile        = "sl-usage-limits-cache.json"
	usageCacheTTLPrompt   = 60 * time.Second
	usageCacheTTLDefault  = 300 * time.Second
	usageAPIURL           = "https://api.anthropic.com/api/oauth/usage"
	usageHTTPTimeout      = 10 * time.Second
)

type usageCache struct {
	Timestamp int64       `json:"timestamp"`
	Status    string      `json:"status"`
	Data      interface{} `json:"data,omitempty"`
}

func runUsageAwareness(cmd *cobra.Command, args []string) error {
	if !config.IsHookEnabled("usage-context-awareness") {
		os.Exit(0)
	}

	// Always output continue:true regardless of errors
	defer func() {
		fmt.Fprint(os.Stdout, `{"continue":true}`)
	}()

	var inputData map[string]interface{}
	if err := json.NewDecoder(os.Stdin).Decode(&inputData); err != nil {
		inputData = map[string]interface{}{}
	}

	_, isUserPrompt := inputData["prompt"].(string)

	if usageShouldFetch(isUserPrompt) {
		fetchAndCacheUsage()
	}

	return nil
}

func usageCachePath() string {
	return filepath.Join(os.TempDir(), usageCacheFile)
}

func usageShouldFetch(isUserPrompt bool) bool {
	ttl := usageCacheTTLDefault
	if isUserPrompt {
		ttl = usageCacheTTLPrompt
	}
	data, err := os.ReadFile(usageCachePath())
	if err != nil {
		return true
	}
	var cache usageCache
	if err := json.Unmarshal(data, &cache); err != nil {
		return true
	}
	age := time.Duration(time.Now().UnixMilli()-cache.Timestamp) * time.Millisecond
	return age >= ttl
}

func writeUsageCache(status string, data interface{}) {
	entry := usageCache{
		Timestamp: time.Now().UnixMilli(),
		Status:    status,
		Data:      data,
	}
	b, err := json.Marshal(entry)
	if err != nil {
		return
	}
	_ = os.WriteFile(usageCachePath(), b, 0644)
}

func getClaudeCredentials() string {
	if runtime.GOOS == "darwin" {
		out, err := exec.Command("security", "find-generic-password", "-s", "Claude Code-credentials", "-w").Output()
		if err == nil {
			raw := strings.TrimSpace(string(out))
			var parsed struct {
				ClaudeAiOauth struct {
					AccessToken string `json:"accessToken"`
				} `json:"claudeAiOauth"`
			}
			if json.Unmarshal([]byte(raw), &parsed) == nil && parsed.ClaudeAiOauth.AccessToken != "" {
				return parsed.ClaudeAiOauth.AccessToken
			}
		}
	}

	home, _ := os.UserHomeDir()
	credPath := filepath.Join(home, ".claude", ".credentials.json")
	data, err := os.ReadFile(credPath)
	if err != nil {
		return ""
	}
	var creds struct {
		ClaudeAiOauth struct {
			AccessToken string `json:"accessToken"`
		} `json:"claudeAiOauth"`
	}
	if err := json.Unmarshal(data, &creds); err != nil {
		return ""
	}
	return creds.ClaudeAiOauth.AccessToken
}

func fetchAndCacheUsage() {
	token := getClaudeCredentials()
	if token == "" {
		writeUsageCache("unavailable", nil)
		return
	}

	client := &http.Client{Timeout: usageHTTPTimeout}
	req, err := http.NewRequest(http.MethodGet, usageAPIURL, nil)
	if err != nil {
		writeUsageCache("unavailable", nil)
		return
	}
	req.Header.Set("Authorization", "Bearer "+token)
	req.Header.Set("anthropic-beta", "oauth-2025-04-20")
	req.Header.Set("User-Agent", "solon/1.0")
	req.Header.Set("Accept", "application/json")

	resp, err := client.Do(req)
	if err != nil {
		writeUsageCache("unavailable", nil)
		return
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		writeUsageCache("unavailable", nil)
		return
	}

	var respData interface{}
	if err := json.NewDecoder(resp.Body).Decode(&respData); err != nil {
		writeUsageCache("unavailable", nil)
		return
	}
	writeUsageCache("available", respData)
}

File: crates/solon-core/src/hooks/usage_awareness.rs
/// UserPromptSubmit/PostToolUse hook: fetch and cache Anthropic usage limits.
/// Always outputs {"continue":true}. Fetches API in background when cache is stale.
use anyhow::Result;
use serde_json::Value;
use std::time::{SystemTime, UNIX_EPOCH};

use crate::hooks::is_hook_enabled;

const USAGE_CACHE_FILE: &str = "sl-usage-limits-cache.json";
const CACHE_TTL_PROMPT_MS: u64 = 60_000; // 1 minute for user prompts
const CACHE_TTL_DEFAULT_MS: u64 = 300_000; // 5 minutes otherwise
const USAGE_API_URL: &str = "https://api.anthropic.com/api/oauth/usage";

pub fn run() -> Result<()> {
    if !is_hook_enabled("usage-context-awareness") {
        std::process::exit(0);
    }

    // Always output continue:true regardless of errors
    let _guard = ContinueGuard;

    let input: Value = crate::hooks::read_hook_input().unwrap_or(Value::Null);
    let is_user_prompt = input.get("prompt").and_then(|v| v.as_str()).is_some();

    if should_fetch(is_user_prompt) {
        fetch_and_cache_usage();
    }

    Ok(())
}

/// RAII guard that always prints {"continue":true} on drop.
struct ContinueGuard;
impl Drop for ContinueGuard {
    fn drop(&mut self) {
        print!("{{\"continue\":true}}");
    }
}

fn cache_path() -> std::path::PathBuf {
    std::env::temp_dir().join(USAGE_CACHE_FILE)
}

fn now_ms() -> u64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default()
        .as_millis() as u64
}

fn should_fetch(is_user_prompt: bool) -> bool {
    let ttl = if is_user_prompt {
        CACHE_TTL_PROMPT_MS
    } else {
        CACHE_TTL_DEFAULT_MS
    };
    let data = match std::fs::read_to_string(cache_path()) {
        Ok(d) => d,
        Err(_) => return true,
    };
    let v: Value = match serde_json::from_str(&data) {
        Ok(v) => v,
        Err(_) => return true,
    };
    let ts = v.get("timestamp").and_then(|t| t.as_u64()).unwrap_or(0);
    now_ms().saturating_sub(ts) >= ttl
}

fn write_cache(status: &str, data: Option<Value>) {
    let entry = serde_json::json!({
        "timestamp": now_ms(),
        "status": status,
        "data": data
    });
    let _ = std::fs::write(
        cache_path(),
        serde_json::to_string(&entry).unwrap_or_default(),
    );
}

fn get_credentials() -> Option<String> {
    // Try ~/.claude/.credentials.json
    let home = dirs::home_dir()?;
    let cred_path = home.join(".claude").join(".credentials.json");
    let data = std::fs::read_to_string(&cred_path).ok()?;
    let v: Value = serde_json::from_str(&data).ok()?;
    let token = v.get("claudeAiOauth")?.get("accessToken")?.as_str()?;
    if token.is_empty() {
        return None;
    }
    Some(token.to_string())
}

fn fetch_and_cache_usage() {
    let token = match get_credentials() {
        Some(t) => t,
        None => {
            write_cache("unavailable", None);
            return;
        }
    };

    // Use std::process to call curl — avoids reqwest/tokio dep
    let output = std::process::Command::new("curl")
        .args([
            "-s",
            "-m",
            "10",
            "-H",
            &format!("Authorization: Bearer {}", token),
            "-H",
            "anthropic-beta: oauth-2025-04-20",
            "-H",
            "User-Agent: solon/1.0",
            "-H",
            "Accept: application/json",
            USAGE_API_URL,
        ])
        .output();

    match output {
        Ok(o) if o.status.success() => {
            let body = String::from_utf8_lossy(&o.stdout);
            if let Ok(data) = serde_json::from_str::<Value>(&body) {
                write_cache("available", Some(data));
            } else {
                write_cache("unavailable", None);
            }
        }
        _ => write_cache("unavailable", None),
    }
}
